design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,385 @@
+2. Deploy 1 Secrets Provider per app / (group of apps), giving the Secrets Provider only access to the secrets they have been given access to in Conjur. Should the user require further separation of duties on Conjur secrets, they will need to deploy another Secrets Provider
+
+The Secrets Provider host will resemble the following:
+
+```yaml
+- !policy
+  id: secrets
+  body:
+  - !host
+    id: secrets-provider-1
+    annotations:
+      authn-k8s/namespace: prod-namespace
+      authn-k8s/service-account: prod-sa
+      authn-k8s/authenticator-container-name: cyberark-secrets-provider
+
+  - !layer 
+
+  - !permit
+    resources: !variable super_secret
+    role: !layer secrets-accessors
+    privileges: [ read, execute ]
+ 
+- !grant
+  role: !layer secrets-accessors
+  members:
+    - !host conjur/authn-k8s/secrets-provider-1
+```
+
+The Secrets Provider manifest will be defined as the following:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: secrets-provider-job
+spec:
+  template:
+    spec:
+      serviceAccountName: prod-sa
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider-1
+
+# Rest of environment variables for Pod to run 
+```
+
+We recommend that the admin will define "Role" instead of "ClusterRole" to follow the principle of least privilege on resources.
+
+#### Code changes
+
+1. Support Job application identity 
+2. Enhance Batch retrieval
+
+*Support Job application identity*
+
+To authenticate to Conjur, the Secrets Provider application identity (or Resource Restrictions) will be the characteristics of the Pod that it is running in. Currently, we do not have a "Job" application identity granularity so to allow the Secrets Provider the ability authenticate to Conjur, the customer will need to choose between the Namespace or Service Account granularities. If there is a requirement to support a "Job" granularity this can be done by adding code to the authenticator in Conjur.
+
+*Enhance Batch retrieval*
+
+At current, the Secrets Provider code looks at the "K8S_SECRET" environment variable in each of the Pods and preforms a batch retrieval against the Conjur API to get the relevant Conjur Secrets. If there is a failure in retrieving one of the secrets, the request fails and no secrets are returned to the request Pod. 
+
+This is problematic for us at this stage because now the Secrets Provider is sitting separately and will need to perform a batch retrieval of all the secrets in the namespace. In other words, if the Secrets Provider fails to fetch one Conjur secret, the whole request will fail and applications will not get their secrets. To avoid this, we will parse over each K8s Secret and perform a batch retrieval request. That way, if there is a failure in retrieving a Conjur secret (permissioning for example), it will be contained to that secret and not impact all the secrets in the namespace.
```

Comment:
can you give an example of how a failed batch-retrieval response looks like? does it state which variables failed to be retrieved?

Also - it seems that this change will affect the performance of the secrets-provider. if the customer defined 1000 k8s secrets that needs Conjur secrets then instead of sending 1 batch-retrieval request to conjur it will send 1000 requests.

How about the following approach - send a batch retrieval with all the required Conjur secrets. If it failed (and the response shows the failed secrets) remove the failing ones from the request and send another request that will pass.  Then you can update all the k8s secrets with the correct value and print a warning (or an error) that some secrets failed to be updated. This way you at least update those that you can. (you will need to investigate the batch response).